src/mlpack/core/hpt/hpt_impl.hpp

```diff
@@ -0,0 +1,246 @@
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         template<typename, typename, typename, typename, typename> class CV,
+         typename Optimizer,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<size_t I, typename ArgsTuple, typename... Collections, typename>
+void HyperParameterTuner<MLAlgorithm,
+                         Metric,
+                         CV,
+                         Optimizer,
+                         MatType,
+                         PredictionsType,
+                         WeightsType>::InitGridSearch(
+    const ArgsTuple& /* args */,
+    Collections... collections)
+{
+  optimizer = Optimizer(collections...);
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         template<typename, typename, typename, typename, typename> class CV,
+         typename Optimizer,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<size_t I, class ArgsTuple, class... Collections, typename, typename>
+void HyperParameterTuner<MLAlgorithm,
+                         Metric,
+                         CV,
+                         Optimizer,
+                         MatType,
+                         PredictionsType,
+                         WeightsType>::InitGridSearch(
+    const ArgsTuple& args,
+    Collections... collections)
+{
+  InitGridSearch<I + 1>(args, collections...);
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         template<typename, typename, typename, typename, typename> class CV,
+         typename Optimizer,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<size_t I, class ArgsTuple, class... Collections, class, class, class>
+void HyperParameterTuner<MLAlgorithm,
+                         Metric,
+                         CV,
+                         Optimizer,
+                         MatType,
+                         PredictionsType,
+                         WeightsType>::InitGridSearch(
+    const ArgsTuple& args,
+    Collections... collections)
+{
+  InitGridSearch<I + 1>(args, collections..., std::get<I>(args));
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         template<typename, typename, typename, typename, typename> class CV,
+         typename Optimizer,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<size_t I, typename ArgsTuple, typename... BoundArgs, typename>
+void HyperParameterTuner<MLAlgorithm,
+                         Metric,
+                         CV,
+                         Optimizer,
+                         MatType,
+                         PredictionsType,
+                         WeightsType>::InitCVFunctionAndOptimize(
+    const ArgsTuple& /* args */,
+    arma::mat& bestParams,
+    BoundArgs... boundArgs)
```

Comment:
It works in the following way. If a user wants to bind some hyper-parameters to some particular values, he/she uses the `Bind` function for that. This `Bind` function returns objects of the type `PreBoundArg`, which are basically wrapped input arguments. We can think about it like the `Bind` function "marks" arguments that should be bound. In order to be completely "bound", information about argument position should be presented. This information is added by using methods with the name `InitCVFunctionAndOptimize`. Eventually "bound args" are passed to `CVFunction`, for which information about the values and their positions in the argument list is sufficient to bind the hyper-parameters to the values.

When you suggested `Fixed`, did you mean using `FixedArgs` instead of `BoundArgs` and preserving other names as they are now?